File: FamlyAndroidEdition/app/src/main/java/com/mack/john/famly_androidedition/utils/LocationUtils.java
package com.mack.john.famly_androidedition.utils;

import android.Manifest;
import android.annotation.SuppressLint;
import android.content.Context;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.location.Address;
import android.location.Geocoder;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.util.Log;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.gms.maps.CameraUpdate;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.model.BitmapDescriptor;
import com.google.android.gms.maps.model.BitmapDescriptorFactory;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.MarkerOptions;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.mack.john.famly_androidedition.R;
import com.mack.john.famly_androidedition.objects.account.profile.Profile;
import com.mack.john.famly_androidedition.objects.post.Post;

import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public class LocationUtils implements LocationListener {



    // Class properties
    private static final String TAG = "LocationUtils";

    Context mContext;

    private final LocationManager mLocationManager;

    private double mLatitude;
    private double mLongitude;
    
    private double mProfileLatitude;
    private double mProfileLongitude;

    private boolean mRequestingUpdates = false;



    // Constructor
    public LocationUtils(Context context) {
        this.mContext = context;

        // Instantiate location manager and call custom method to get current location
        this.mLocationManager = (LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
        getCurrentLocation();
    }



    // System generated methods
    @Override
    public void onLocationChanged(Location location) {
        // Get current latitude and longitude
        mLatitude = location.getLatitude();
        mLongitude = location.getLongitude();

        // Reference firebase user account and database
        FirebaseUser firebaseAccount = FirebaseAuth.getInstance().getCurrentUser();
        FirebaseFirestore database = FirebaseFirestore.getInstance();

        // Referece current date
        final Date date = new Date();

        // Map location data for upload
        Map<String, Object> checkinMap = new HashMap<>();
        checkinMap.put("latitude", mLatitude);
        checkinMap.put("longitude", mLongitude);
        checkinMap.put("checkinTimestamp", date);

        // Reference database location and attempt to upload
        database.collection("accounts").document(firebaseAccount.getUid()).collection("profiles")
                .document(firebaseAccount.getUid() + AccountUtils.loadProfile(mContext).getProfileId()).collection("location")
                .document("location").set(checkinMap).addOnCompleteListener(new OnCompleteListener<Void>() {
            @Override
            public void onComplete(@NonNull Task<Void> task) {
                if(task.isSuccessful()) {
                    Log.i(TAG, "onComplete: Location Updated Successfully");
                }
            }
        });
    }

    @Override
    public void onStatusChanged(String provider, int status, Bundle extras) {

    }

    @Override
    public void onProviderEnabled(String provider) {

    }

    @Override
    public void onProviderDisabled(String provider) {

    }



    // Custom methods
    // Custom method to get current location
    private void getCurrentLocation() {
        // If required permissions have been granted, continue
        if(ContextCompat.checkSelfPermission(mContext, Manifest.permission.ACCESS_FINE_LOCATION)
                == PackageManager.PERMISSION_GRANTED) {

            // Start requesting location updates
            mLocationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, 300000, 10.0f, this);
            mRequestingUpdates = true;

            // Pull current location
            Location currentPosition = mLocationManager.getLastKnownLocation(LocationManager.GPS_PROVIDER);

            // If current location is available, extract latitude and longitude
            if(currentPosition != null) {
                mLatitude = currentPosition.getLatitude();
                mLongitude = currentPosition.getLongitude();
            }
        }
    }

    // Custom method to get profile's last known location
    public void getProfileLocation(final Profile profile, final GoogleMap map) {

        // Reference firebase user accounnt and database
        FirebaseUser firebaseAccount = FirebaseAuth.getInstance().getCurrentUser();
        FirebaseFirestore database = FirebaseFirestore.getInstance();

        // Reference database location and attempt to download
        database.collection("accounts").document(firebaseAccount.getUid()).collection("profiles")
                .document(firebaseAccount.getUid() + profile.getProfileId()).collection("location")
                .document("location").get().addOnCompleteListener(new OnCompleteListener<DocumentSnapshot>() {
            @Override
            public void onComplete(@NonNull Task<DocumentSnapshot> task) {

                // If successful...
                if(task.isSuccessful()) {

                    // If a map does not exist, do not continue
                    if(map == null) {
                        return;
                    }

                    DocumentSnapshot document = task.getResult();

                    // If location information exists...
                    if(document.exists()) {

                        // Reference location data
                        double latitude = document.getDouble("latitude");
                        double longitude = document.getDouble("longitude");

                        // Reference current latitude and longitude
                        LatLng currentLocation = new LatLng(latitude, longitude);

                        // Zoom camera in to current location
                        CameraUpdate cameraMovement = CameraUpdateFactory.newLatLngZoom(currentLocation, 15);
                        map.animateCamera(cameraMovement);

                        // Reference marker custom color
                        @SuppressLint("ResourceType")
                        String green = mContext.getResources().getString(R.color.colorGrassPrimary);

                        // Create marker and set title / snippet
                        MarkerOptions options = new MarkerOptions();
                        options.title(AccountUtils.loadProfile(mContext).getFirstName() + " " + AccountUtils.loadAccount(mContext).getFamilyName());
                        options.snippet(convertAddress(mContext, latitude, longitude));

                        // Set marker location and custom color
                        options.position(currentLocation);
                        options.icon(setMarkerColor(green));

                        // Add marker to map
                        map.addMarker(options);
                    }

                    // If there is a problem, let the user know
                    else {
                        Toast.makeText(mContext, mContext.getString(R.string.location_info_for) + " " + profile.getFirstName() + " "
                                + mContext.getString(R.string.is_unavailable), Toast.LENGTH_SHORT).show();
                    }
                }

                // If there is a problem, let the user know
                else {
                    Toast.makeText(mContext, mContext.getString(R.string.location_info_for) + " " + profile.getFirstName() + " "
                            + mContext.getString(R.string.is_unavailable), Toast.LENGTH_SHORT).show();
                }
            }
        });
    }

    // Custom method to return current location latitude
    public double getLatitude() {
        return mLatitude;
    }

    // Custom method to return current location longitude
    public double getLongitude() {
        return mLongitude;
    }

    // Custom method to create new checkin
    public void checkIn(final Context context, final double latitude, final double longitude, final String message, final Bitmap image) {

        // Reference firebase user account and database
        FirebaseUser firebaseAccount = FirebaseAuth.getInstance().getCurrentUser();
        FirebaseFirestore database = FirebaseFirestore.getInstance();

        // Reference current date
        final Date date = new Date();

        // Map checkin for upload
        Map<String, Object> checkinMap = new HashMap<>();
        checkinMap.put("latitude", latitude);
        checkinMap.put("longitude", longitude);
        checkinMap.put("checkinTimestamp", date);

        // Reference databse location and attempt to upload
        database.collection("accounts").document(firebaseAccount.getUid()).collection("profiles")
                .document(firebaseAccount.getUid() + AccountUtils.loadProfile(context).getProfileId()).collection("location")
                .document("location").set(checkinMap).addOnCompleteListener(new OnCompleteListener<Void>() {
            @Override
            public void onComplete(@NonNull Task<Void> task) {

                // Convert location to readable address
                String address = convertAddress(context, latitude, longitude);

                // Load logged in profile
                Profile profile = AccountUtils.loadProfile(context);

                // Indicate whether or not checkin has an image
                boolean hasImage;
                if(image != null) {
                    hasImage = true;
                }

                else {
                    hasImage = false;
                }

                // If checkin has a message, add it to the post
                if(message == null || message.trim().equals("")) {
                    String postMesage = AccountUtils.loadProfile(context).getFirstName() + " " + context.getString(R.string.checked_in_at) + " " + address;

                    Post post = new Post(postMesage, date, profile.getProfileId(), profile.getFirstName(), hasImage);
                    PostUtils.createPost(context, post, profile, image);
                }

                // Otherwise, post without a message
                else {
                    String postMessage = AccountUtils.loadProfile(context).getFirstName() + " " + context.getString(R.string.checked_in_at) + " " + address
                            + " " + context.getString(R.string.and_says) + " " + message + "'";

                    Post post = new Post(postMessage, date, profile.getProfileId(), profile.getFirstName(), hasImage);
                    PostUtils.createPost(context, post, profile, image);
                }
            }
        });
    }

    // Custom method to update location display for current profile
    public void updateLocationDisplay(final Context context, View view, final Profile profile) {

        // Reference UI elements
        final TextView timeStampDisplay = view.findViewById(R.id.display_timestamp);
        final TextView lastKnownLocationDisplay = view.findViewById(R.id.display_last_location);

        // Cleaar UI elements by default
        timeStampDisplay.setText("");
        lastKnownLocationDisplay.setText("");

        // Reference firebase user account and databse
        FirebaseUser firebaseAccount = FirebaseAuth.getInstance().getCurrentUser();
        FirebaseFirestore database = FirebaseFirestore.getInstance();

        // Reference database location and attempt to download
        database.collection("accounts").document(firebaseAccount.getUid()).collection("profiles")
                .document(firebaseAccount.getUid() + profile.getProfileId()).collection("location")
                .document("location").get().addOnCompleteListener(new OnCompleteListener<DocumentSnapshot>() {
            @Override
            public void onComplete(@NonNull Task<DocumentSnapshot> task) {

                // If successful...
                if(task.isSuccessful()) {
                    DocumentSnapshot document = task.getResult();

                    // If data is available...
                    if(document.exists()) {

                        // Reference location data
                        Date checkinTimestamp = document.getDate("checkinTimestamp");
                        double latitude = document.getDouble("latitude");
                        double longitude = document.getDouble("longitude");
                        String address = convertAddress(context, latitude, longitude);

                        // Format timestamp
                        SimpleDateFormat dateFormat = new SimpleDateFormat("MMMM dd, yyyy @ hh:mm a", Locale.getDefault());
                        String timestamp = dateFormat.format(checkinTimestamp);

                        // Display location data
                        timeStampDisplay.setText(timestamp);
                        lastKnownLocationDisplay.setText(address);
                    }

                    // If there is a problem, let the user know
                    else {
                        Toast.makeText(mContext, mContext.getString(R.string.location_info_for) + " " + profile.getFirstName() + " "
                                + mContext.getString(R.string.is_unavailable), Toast.LENGTH_SHORT).show();
                        
                        timeStampDisplay.setText("");
                        lastKnownLocationDisplay.setText(context.getString(R.string.unknown));
                    }
                }

                // If there is a problem, let the user know
                else {
                    Toast.makeText(mContext, mContext.getString(R.string.location_info_for) + " " + profile.getFirstName() + " "
                            + mContext.getString(R.string.is_unavailable), Toast.LENGTH_SHORT).show();
                }
            }
        });
    }

    // Custom method to convert location to readable address
    public String convertAddress(Context context, double latitude, double longitude) {

        // Set address as unknown by default
        String addressString = "Unknown Address";

        try {

            // Convert location to address
            Geocoder geocoder = new Geocoder(context, Locale.getDefault());
            List<Address> addressList = geocoder.getFromLocation(latitude, longitude, 1);
            Address address = addressList.get(0);
            String streetAddress = address.getAddressLine(0);
            addressString = streetAddress;
        }

        catch (IOException e) {
            e.printStackTrace();
        }

        // Return address
        return addressString;
    }

    // Custom method to set map marker color
    private BitmapDescriptor setMarkerColor(String color) {
        float[] hsv = new float[3];
        Color.colorToHSV(Color.parseColor(color), hsv);
        return BitmapDescriptorFactory.defaultMarker(hsv[0]);
    }
}

File: Fam.ly - iOS Edition/Fam.ly - iOS Edition/Utils/LocationUtils.swift
//
//  LocationUtils.swift
//  Fam.ly - iOS Edition
//
//  Created by Johnny Mack on 6/17/19.
//  Copyright © 2019 John Mack. All rights reserved.
//

import Foundation
import GoogleMaps
import FirebaseFirestore
import FirebaseAuth

class LocationUtils {
    
    
    
    // Class properties
    let mLocationManager = CLLocationManager()
    
    var mLatitude: Double?
    var mLongitude: Double?
    
    
    
    // Connstructor
    init(Location location: CLLocation?) {
        if location != nil {
            self.mLatitude = location!.coordinate.latitude
            self.mLongitude = location!.coordinate.longitude
        }
    }
    
    
    
    // Custom methods
    // Custom method to update current user's location
    public func updateLocation(Parent parent: Parent?, Child child: Child?) {
        
        // Reference profile id
        var profileId: String?
        
        if parent != nil {
            profileId = parent?.getProfileId()
        }
        
        else if child != nil {
            profileId = child?.getProfileId()
        }
        
        // Reference database
        let database = Firestore.firestore()
        
        // Reference current date
        let date = Date()
        
        // Map checkin for upload
        let checkinData: [String: Any] = [
            "latitude": mLatitude ?? 0,
            "longitude": mLongitude ?? 0,
            "checkinTimestamp": date
        ]
        
        // Reference database location and attempt to upload
        database.collection("accounts").document(Auth.auth().currentUser!.uid).collection("profiles").document(Auth.auth().currentUser!.uid + profileId!).collection("location").document("location").setData(checkinData) { err in
            if let err = err {
                print("Error writing document: \(err)")
            }
            
            else {
                print("Location updated")
            }
        }
    }
    
    // Custom method to get profile's last known location
    public func getProfileLocation(Controller controller: UIViewController, Parent parent: Parent?, Child child: Child?, Map map: GMSMapView) {
        
        // Reference profile id and name
        var profileId: String?
        var firstName: String?
        
        if parent != nil {
            profileId = parent?.getProfileId()
            firstName = parent?.getFirstName()
        }
        
        else if child != nil {
            profileId = child?.getProfileId()
            firstName = child?.getFirstName()
        }
        
        // Reference database location and attempt to download
        let database = Firestore.firestore()
        let documentRef = database.collection("accounts").document(Auth.auth().currentUser!.uid).collection("profiles").document(Auth.auth().currentUser!.uid + profileId!).collection("location").document("location")
        
        documentRef.getDocument { (document, error) in
            
            // If documennt exists...
            if let document = document, document.exists {
                
                // Get latitude and longitude
                self.mLatitude = document.get("latitude") as? Double
                self.mLongitude = document.get("longitude") as? Double
                
                // Convert to location data
                let location = CLLocation(latitude: self.mLatitude!, longitude: self.mLongitude!)
                
                // Show location on map
                map.camera = GMSCameraPosition(target: location.coordinate, zoom: 15, bearing: 0, viewingAngle: 0)
                
                // Conver location to readable address
                let geocoder = GMSGeocoder()
                geocoder.reverseGeocodeCoordinate(location.coordinate) { (response, error) in
                    guard let address = response?.firstResult(), let lines = address.lines else {
                        return
                    }
                    
                    // Add marker to map
                    let marker = GMSMarker()
                    marker.position = location.coordinate
                    marker.title = firstName! + " " + (AccountUtils.loadAccount()?.getFamilyName())!
                    marker.snippet = lines.joined(separator: " ")
                    marker.map = map
                }
            }
            
            // If there was a problem, let the user know
            else {
                let alert = UIAlertController(title: "Unable to Locate", message: "Location information for " + firstName! + " is unavailable.", preferredStyle: .alert)
                alert.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
                controller.present(alert, animated: true)
            }
        }
    }
    
    // Custom method to create new checkin
    public func checkIn(Controller controller: UIViewController, Parent parent: Parent?, Child child: Child?, Message message: String?, Image image: UIImage?) {
        
        // Reference profile id and name
        var profileId: String?
        var firstName: String?
        
        if parent != nil {
            profileId = parent?.getProfileId()
            firstName = parent?.getFirstName()
        }
        
        else if child != nil {
            profileId = child?.getProfileId()
            firstName = child?.getFirstName()
        }
        
        // Reference database
        let database = Firestore.firestore()
        
        // Reference current date
        let date = Date()
        
        // Map checkin for upload
        let checkinData: [String: Any] = [
            "latitude": mLatitude!,
            "longitude": mLongitude!,
            "checkinTimestamp": date
        ]
        
        // Reference database location and attempt to upload
        database.collection("accounts").document(Auth.auth().currentUser!.uid).collection("profiles").document(Auth.auth().currentUser!.uid + profileId!).collection("location").document("location").setData(checkinData) { err in
            
            // If there was a problem, let the user know
            if let err = err {
                print("Error writing document: \(err)")
                
                let alert = UIAlertController(title: "Uh Oh!", message: "There was a problem checking in.", preferredStyle: .alert)
                alert.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
                controller.present(alert, animated: true)
            }
                
            // If successful...
            else {
                
                // Create location data
                let location = CLLocation(latitude: self.mLatitude!, longitude: self.mLongitude!)
                
                // Convert location to readable address
                let geocoder = GMSGeocoder()
                geocoder.reverseGeocodeCoordinate(location.coordinate) { (response, error) in
                    guard let address = response?.firstResult(), let lines = address.lines else {
                        return
                    }
                    
                    // Indicate whether or not post includes a photo
                    var hasImage = false
                    
                    if image != nil {
                        hasImage = true
                    }
                    
                    else {
                        hasImage = false
                    }
                    
                    // If post has a message include it and create newsfeed post
                    if message == nil || message?.trimmingCharacters(in: .whitespacesAndNewlines) == "" {
                        let postMessage = firstName! + " checked in at " + lines.joined(separator: " ")
                        
                        let post = Post(PostMessage: postMessage, TimeStamp: date, PosterId: profileId!, PosterName: firstName!, HasImage: hasImage)
                        PostUtils.createPost(Post: post, Controller: controller, Photo: image)
                    }
                    
                    // Otherwise, create post without a message
                    else {
                        let postMessage = firstName! + " checked in at " + address.lines!.joined(separator: " ") + " and says '" + message! + "'"
                        
                        let post = Post(PostMessage: postMessage, TimeStamp: date, PosterId: profileId!, PosterName: firstName!, HasImage: hasImage)
                        PostUtils.createPost(Post: post, Controller: controller, Photo: image)
                    }
                }
            }
        }
    }
    
    // Custom method to update location display for current profile
    public func updateLocationDisplay(Controller controller: UIViewController, TimeStampDisplay timeStampDisplay: UILabel, LastKnowLocationDisplay lastKnownLocationDisplay: UILabel, Parent parent: Parent?, Child child: Child?) {
        
        // Clear UI elements
        timeStampDisplay.text = ""
        lastKnownLocationDisplay.text = ""
        
        // Reference profile id and name
        var profileId: String?
        var firstName: String?
        
        if parent != nil {
            profileId = parent?.getProfileId()
            firstName = parent?.getFirstName()
        }
        
        else if child != nil {
            profileId = child?.getProfileId()
            firstName = child?.getFirstName()
        }
        
        // Reference database location and attempt to download
        let database = Firestore.firestore()
        let documentRef = database.collection("accounts").document(Auth.auth().currentUser!.uid).collection("profiles").document(Auth.auth().currentUser!.uid + profileId!).collection("location").document("location")
        
        documentRef.getDocument { (document, error) in
            
            // If successful...
            if let document = document, document.exists {
                
                // Reference post data
                self.mLatitude = document.get("latitude") as? Double
                self.mLongitude = document.get("longitude") as? Double
                let timeStamp = document.get("checkinTimestamp") as? Date
                
                // Create location
                let location = CLLocation(latitude: self.mLatitude!, longitude: self.mLongitude!)
                
                // Convert location to readable address
                let geocoder = GMSGeocoder()
                geocoder.reverseGeocodeCoordinate(location.coordinate) { (response, error) in
                    guard let address = response?.firstResult(), let lines = address.lines else {
                        return
                    }
                    
                    // Display timestamp
                    let dateFormatter = DateFormatter()
                    dateFormatter.dateFormat = "MMMM dd, yyyy @ hh:mm a"
                    let dateString = dateFormatter.string(from: timeStamp!)
                    
                    // Display address
                    timeStampDisplay.text = dateString
                    lastKnownLocationDisplay.text = lines.joined(separator: " ")
                }
            }
            
            // If there was a problem, let the user know
            else {
                lastKnownLocationDisplay.text = "Unknown Location"
                let alert = UIAlertController(title: "Unable to Locate", message: "Location information for " + firstName! + " is unavailable.", preferredStyle: .alert)
                alert.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
                controller.present(alert, animated: true)
            }
        }
    }
}
